File: akbarsh/Wildcard.py
"""
Input: 10?
Output: 101, 100
i.e. ? behaves like a wild-card. There are two possibilities for 10?, when that ? is replaced with
either 0 or 1.
Input: 1?0?
Output: 1000, 1001, 1100, 1101
Please write a program that takes given strings as input and produces the suggested output.
"""
def wildcard(str):
    st=0
    str=list(str)
    return wildcard_helper(str,st)
def wildcard_helper(str,st):
    if st>=len(str):
        print ''.join(str)
        return
    elif str[st]=='?':
        str[st]='0'
        wildcard_helper(str,st+1)
        str[st] = '1'
        wildcard_helper(str, st + 1)
        str[st]='?' #we need to backtack as string is passed by ref
    else:
        wildcard_helper(str, st + 1)
wildcard('10?')
wildcard('10??1')

File: ahmedhasan/recursion/WildCard.java

public class WildCard {
	public static void main(String[] args) {
        WildCard app = new WildCard();
        app.wildCard("1?0?");
    }
	
	
	    public void wildCard(String input) {
	        wildCardHelper(input,"");
	    }
	    private void wildCardHelper(String input,String soFar) {
	        if (input.length() == 0)
	            System.out.println(soFar);
	        else {
	            if (input.charAt(0) == '?') {
	                wildCardHelper(input.substring(1),soFar + '0');
	                wildCardHelper(input.substring(1),soFar + '1');
	            }
	            else
	                wildCardHelper(input.substring(1),soFar+input.charAt(0));
	        }

	    }

	    
	}
